### ISSUE TITLE
Append label mask to input

### ISSUE THREAD

**Issue by Karol-G** [MIC-DKFZ/nnUNet#430]
Hi Fabian,

I want to append a mask with labels to my input in the form of an additional channel to guide the model during training.
So when the data is being augmented flips, elastic deformation and so on should be apllied to the mask too but not augmentations like brightness changes. 
If I am not mistaken this is already implemented with "seg channels", but I am not sure if I understood it correctly.
I believe you talked about this in #268 and #415 already too.

Is this implemented already and if yes, can you tell me roughly what I need to do?

Best
Karol

**Comment by FabianIsensee** [comment]
Hi Karol,
you need to place this additional segmentation in the 'seg' key:value pair because you don't want to have brightness, Gaussian noise etc to be applied to it. At the end of the pipeline, you need to move it to 'data' which you can do with MoveSegAsOneHotToData (I recommend one hot encoding for this). See also here:
https://github.com/MIC-DKFZ/nnUNet/blob/e26f6de358d15ccf66c55180a63d9cb91f0d2500/nnunet/training/data_augmentation/default_data_augmentation.py#L175

The augmentations of the cascade do exactly that. You can use them as inspiration
Best,
Fabian

**Comment by Karol-G** [comment]
Hi Fabian,

thanks for the quick answer! I wil test this out as soon as I can :)
Will close this for now then.

**Comment by Karol-G** [comment]
Hi again,

i kinda forgot about this, but now I have time to finally work on this.
I have looked into the code now and my approach would be the following:

1. Append the masks for guidance to each image as an additional channel in the raw unprocessed dataset
2. Perform `nnUNet_plan_and_preprocess`
3. In `dataset_loading.py -> DataLoader3D -> generate_train_batch()` add the mask channel to `seg`
4. In `nnUNetTrainerV2.py -> nnUNetTrainerV2 -> setup_DA_params()` set self.data_aug_params['selected_seg_channels'] = [0,1], set self.data_aug_params['move_last_seg_chanel_to_data'] = True and set self.data_aug_params['all_segmentation_labels'] = list(range(1, self.num_classes))

I am not sure if all the steps are 100% correct but I belive the approach is correct?
Some questions:
- Does adding the mask as channel to the image before the planning and preprocessing step lead to any unwanted side effects during planning and preprocessing? Like that the model depth will be different or something?
- Is step 3/4. correct? I am unsure about the selected_seg_channels indices. Do I set my mask as `seg[j, 1]` like it would be the case for `seg_from_previous_stage` in line 370? Assuming of course I am not doing cascade training.

Best
Karol

**Comment by Karol-G** [comment]
One further question: In `generate_train_batch()` when getting the guiding mask from case_all_data, is my guiding mask the in the last index or in the second last? So is it [input, guiding_mask, ground_truth] or [input, ground_truth, guiding_mask]?
I am currently assuming it is in the second last, but I am not sure.

```python
seg[j, 0] = np.pad(case_all_data[-1:], ((0, 0),
                                        (-min(0, bbox_x_lb), max(bbox_x_ub - shape[0], 0)),
                                        (-min(0, bbox_y_lb), max(bbox_y_ub - shape[1], 0)),
                                        (-min(0, bbox_z_lb), max(bbox_z_ub - shape[2], 0))),
                   'constant', **{'constant_values': -1})

guiding_mask_index = 1
if seg_from_previous_stage is not None:
    guiding_mask_index = 2

seg[j, guiding_mask_index] = np.pad(case_all_data[-2:-1], ((0, 0),
                                          (-min(0, bbox_x_lb), max(bbox_x_ub - shape[0], 0)),
                                          (-min(0, bbox_y_lb), max(bbox_y_ub - shape[1], 0)),
                                          (-min(0, bbox_z_lb), max(bbox_z_ub - shape[2], 0))),
                   'constant', **{'constant_values': -1})
```

Best
Karol

**Comment by FabianIsensee** [comment]
Hi,
what you are trying to do is something I have not attempted before so there are a lot of pitfalls you could encounter. I think it makes sense to just go through this step by step with some python debugging tool and verify that everything works as intended. It always makes sense to visualize the data with some sort of tool (for example my [BatchViever](https://github.com/FabianIsensee/BatchViewer) project).
I cannot give you a full solution on how to achieve what you are looking for. I do not have any time at all at the moment.
One tip I would like to give you though is that you need to think about the preprocessing: 1) you do not want to apply any form of intensity normalization to the segmentation and 2) you need to take special care with your mask during resampling. Both changes require you to add custom code to the preprocessing pipeline. For 1), hava a look at how ct images are handled. You should add a new CT-like exception for your mask that tells the proprocessor not to run any preprocessing on this input. For 2, you need to ensure you are setting the seg=True flag for resampling that one channel. Both of these can be addressed in 
https://github.com/MIC-DKFZ/nnUNet/blob/eb44c39434ac6c9d2873a5a556e1fb56326a3aa6/nnunet/preprocessing/preprocessing.py#L225

(you also need to modify https://github.com/MIC-DKFZ/nnUNet/blob/eb44c39434ac6c9d2873a5a556e1fb56326a3aa6/nnunet/experiment_planning/experiment_planner_baseline_3DUNet.py#L359)

Best,
Fabian
